File: Examples/typescript/basics.ts
interface Shape {
  area(): number;
  name: string;
}

type Pair<T> = { first: T; second: T };

class Circle implements Shape {
  name: string = "circle";
  constructor(private r: number) {}
  area(): number { return 3.14159 * this.r * this.r; }
}

class Rect implements Shape {
  name = "rect";
  constructor(public w: number, public h: number) {}
  area(): number { return this.w * this.h; }
}

enum Color { Red = 1, Green, Blue }

function total<T extends Shape>(shapes: T[]): number {
  return shapes.reduce((sum: number, s: T) => sum + s.area(), 0);
}

const shapes: Shape[] = [new Circle(2), new Rect(3, 4)];
for (const s of shapes) {
  console.log(s.name, s.area().toFixed(3));
}
console.log(total(shapes).toFixed(2));

const p: Pair<number> = { first: 1, second: 2 };
console.log(p.first + p.second);

const names: string[] = ["bob", "alice"];
names.sort();
console.log(names);

let count: number = 0;
const inc = (by: number = 1): number => (count += by);
inc();
inc(5);
console.log(count);

function greet(who: string, greeting?: string): string {
  return `${greeting ?? "Hello"}, ${who}!`;
}
console.log(greet("world"));
console.log(greet("world", "Hi"));

const m: Map<string, number> = new Map();
m.set("a", 1);
console.log(m.get("a"), m.size);

console.log(Color.Red, Color.Green, Color.Blue);

File: Examples/nim/basics.nim
import strutils, tables

proc fib(n: int): int =
  if n < 2:
    return n
  return fib(n - 1) + fib(n - 2)

proc sum(xs: seq[int]): int =
  for x in xs:
    result = result + x

type
  Color = enum
    Red, Green, Blue
  Point = object
    x, y: int
  Shape = ref object of RootObj
    name: string
  Circle = ref object of Shape
    r: float

method area(s: Shape): float =
  0.0

method area(c: Circle): float =
  3.14159 * c.r * c.r

proc `$`(p: Point): string =
  "(" & $p.x & ", " & $p.y & ")"

var results: seq[int] = @[]
for i in 0..9:
  results.add(fib(i))
echo "fib: ", results

let xs = @[5, 3, 9, 1, 7]
echo "sum = ", sum(xs)
echo "len = ", xs.len
echo "max = ", max(xs)

var p = Point(x: 3, y: 4)
echo "point: ", $p
echo "shifted: ", p.x + 1

let c = Circle(name: "c1", r: 2.0)
echo "area = ", c.area()

var counts = initTable[string, int]()
for word in "a b a c b a".split(" "):
  if counts.hasKey(word):
    counts[word] = counts[word] + 1
  else:
    counts[word] = 1
echo "a=", counts["a"], " b=", counts["b"], " c=", counts["c"]

var i = 0
while i < 3:
  echo "i=", i
  i = i + 1

case 7
of 1:
  echo "one"
of 7:
  echo "seven"
else:
  echo "other"

let color = Green
case color
of Red:
  echo "red"
of Green:
  echo "green"
else:
  echo "blue"

try:
  discard parseInt("abc")
except ValueError as e:
  echo "caught: ", e.msg

echo "10 div 3 = ", 10 div 3
echo "10 mod 3 = ", 10 mod 3
echo "upper: ", toUpperAscii("hello")
echo "joined: ", @["a", "b", "c"].join("-")
